### PR TITLE
TZC380 autoconfiguration (search correct region size) and support for i.MX6Q/D

### PR DESCRIPTION
**Pull request by Emantor**
Hello,

this PR adds a new function called `tzc380_auto_configure` which searches for the matching configuration for a given start address and size. This configuration is then applied with the given permissions to the controller.
This vastly simplifies configurations for devices with a generic RAM layout, since the necessary information is already available from the header file.
Support for i.MX6QD devices is added as a user of the new function.

### CODE REVIEW

File: core/arch/arm/plat-imx/drivers/tzasc.c [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * All rights reserved.
+ *
+ * Rouven Czerwinski <entwicklung@pengutronix.de>
+ */
+
+#include <assert.h>
+#include <drivers/tzc380.h>
+#include <imx-regs.h>
+#include <imx.h>
+#include <initcall.h>
```
Comment (Emantor) [inline core/arch/arm/plat-imx/drivers/tzasc.c]: unused include

File: core/arch/arm/plat-imx/drivers/tzasc.c [hunk 2]
```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2019 Pengutronix
+ * All rights reserved.
+ *
+ * Rouven Czerwinski <entwicklung@pengutronix.de>
+ */
+
+#include <assert.h>
```
Comment (Emantor) [inline core/arch/arm/plat-imx/drivers/tzasc.c]: unused include

File: core/arch/arm/plat-imx/drivers/tzasc.c [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+#include <imx-regs.h>
+#include <imx.h>
+#include <initcall.h>
+#include <mm/core_memprot.h>
+#include <mm/generic_ram_layout.h>
+
+#if defined(CFG_MX6Q) || defined(CFG_MX6D)
+void imx_configure_tzasc(void)
+{
+
```
Comment (etienne-lms) [inline core/arch/arm/plat-imx/drivers/tzasc.c]: remove empty line

File: core/arch/arm/plat-imx/imx6.c [hunk 4]
```diff
@@ -36,6 +36,7 @@
 #include <mm/core_memprot.h>
 #include <mm/core_mmu.h>
 #include <platform_config.h>
+#include <imx.h>
```
Comment (etienne-lms) [inline core/arch/arm/plat-imx/imx6.c]: suggest to move this up.

**Comment by Emantor** [comment]
@etienne-lms Thanks for the review, I addressed your comments and applied the provided tags.
(I just noticed that separate review commits are more common here, I however already rebased and applied the fixups, sorry.)

**Comment by Emantor** [comment]
@MrVan I need your Ack for at least 6e3f9164ff3dce1e3e0fbb7aa9d2cde4ac2fa414
@jforissier, @jenswi-linaro any comments on this PR?

File: core/arch/arm/plat-imx/imx6.c [hunk 5]
```diff
@@ -77,5 +78,8 @@ void plat_cpu_reset_late(void)
 		     addr != CSU_BASE + CSU_CSL_END;
 		     addr += 4)
 			io_setbits32(addr, CSU_SETTING_LOCK);
+#if defined(CFG_TZC380)
```
Comment (jforissier) [inline core/arch/arm/plat-imx/imx6.c]: Conditional not needed (assuming `imx.h` is changed as suggested above)

**Comment by jforissier** [comment]
For "plat-imx: mx6: add support for the TZC380 to MX6Q": LGTM in principle.
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by jforissier** [comment]
Sorry pasted the wrong commit subject above, now fixed.

**Comment by MrVan** [comment]
Acked-by: Peng Fan <peng.fan@nxp.com>

**Comment by Emantor** [comment]
@mrvan for the whole series or for the plat-imx commit?

File: core/arch/arm/plat-imx/imx.h [hunk 6]
```diff
@@ -33,5 +33,10 @@ uint32_t imx_soc_type(void);
 void imx_gpcv2_set_core1_pdn_by_software(void);
 void imx_gpcv2_set_core1_pup_by_software(void);
 void imx_gpcv2_set_core_pgc(bool enable, uint32_t offset);
+
+#if !(defined(CFG_MX6Q) || defined(CFG_MX6D) || defined(CFG_TZC380))
+void imx_configure_tzasc(void) {}
```
Comment (jforissier) [inline core/arch/arm/plat-imx/imx.h]: `static inline void` please.

Minor: it's usually written the other way around, without a negation (if defined() ... else ... endif)

Comment (Emantor) [reply core/arch/arm/plat-imx/imx.h]: I also forgot the `#` for `else`. Fixed with a new Review commit and corrected logic.

File: core/arch/arm/plat-imx/imx.h [hunk 7]
```diff
@@ -34,9 +34,9 @@ void imx_gpcv2_set_core1_pdn_by_software(void);
 void imx_gpcv2_set_core1_pup_by_software(void);
 void imx_gpcv2_set_core_pgc(bool enable, uint32_t offset);
 
-#if !(defined(CFG_MX6Q) || defined(CFG_MX6D) || defined(CFG_TZC380))
-void imx_configure_tzasc(void) {}
-else
+#if (defined(CFG_MX6Q) || defined(CFG_MX6D)) && defined(CFG_TZC380)
```
Comment (etienne-lms) [inline core/arch/arm/plat-imx/imx.h]: `#if defined(CFG_TZC380)` should be enough.

File: core/drivers/tzc380.c [hunk 8]
```diff
@@ -216,6 +216,68 @@ uint32_t tzc_get_action(void)
...
+	uint32_t mask = 0;
+	int i = 0;
+	uint8_t pow = TZC380_POW;
+
+	while (lsize != 0 && pow > 15) {
+		region_size = 1ULL << pow;
+
+		/* Case region fits alignment and covers requested area */
+		if ((address % region_size == 0) &&
+			((address + lsize) % region_size == 0)) {
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/tzc380.c]: please address this minor comment.

[the remark this answers is not in the document]
Comment (Emantor) [reply core/drivers/tzc380.c]: I missed this, should be good now 👍

**Comment by Emantor** [comment]
Squashed down the commits and applied the tags.

**Comment by jbech-linaro** [comment]
Thanks @Emantor , patch merged!

**Comment by Emantor** [comment]
Thanks for the reviews @etienne-lms @jforissier

**Pull request merged**